eliminate emit_expr alternatives

vtjnash opened the pull request:
the different between the several emit_expr alternatives was mostly superficial following the codegen_rewrite PRs. this finishes the work of merging the potential optimizations paths from this set of functions (which was previously a question of whether to return constants as boxed or unboxed values).

`emit_expr` is now the only way to go from `jl_expr_t` -> `jl_cgval_t`.

to go from `jl_cgval_t` to `Value*`, there are a few options:
- `emit_unboxed` -> get the raw bits, unboxed. (occasionally instead use `.V` if `ispointer = false` and you don't want any llvm type conversion to happen)
- `boxed` -> get a `jl_value_t*` (plain and simple). if it needed to box the value, it makes the gcroot also, unless hinted not to by the caller
- `data_pointer` -> get the raw pointer for a pointer value (`ispointer = true`)
- `.V` -> accessing the value field directly is usually discouraged, but occasionally necessary

if you access `.V` or `data_pointer` directly, the code author is responsible for calling `mark_gc_use`, if the usage may not be "obvious" to the compiler (for example, a `CallExpr` usage can't be optimized away, whereas an PtrToInt or GEP may be hard for the root detection algorithm to track)

vtjnash committed:
eliminate emit_expr alternatives

this also implements trivial constant propagation as an improved replacement for the manual emit_unboxed / emit_expr(.., false) optimization

tkelman commented:
Travis failure was a cache corruption, cleared cache for this branch and restarted
```
Makefile:44: Makefile.config: No such file or directory
Makefile:151: /Makefile.rules: No such file or directory
make[3]: *** No rule to make target `/Makefile.rules'.  Stop.
make[3]: Leaving directory `/home/travis/build/JuliaLang/julia/deps/build/llvm-3.7.1/build_Release'
make[2]: *** [/home/travis/build/JuliaLang/julia/usr/lib/libLLVMCodeGen.a] Error 2
```

vtjnash commented:
yep, appveyor failure was a issue (now fixed) where it was trying to make jl_bottom_type into a real ghostValue (instead of undef). unless someone indicates they plan to review, i'll merge this sometime tomorrow.

vtjnash committed:
improve codegen handling of fully linear IR

tkelman commented:
Looks sane to me, though I'm not entirely qualified to judge this code.

vtjnash committed:
constant propogate constant ::Type{T} arguments during codegen

this helps with code-generation of certain intrinsics calls when type inference has not run

vtjnash committed:
compute constant call expressions during codegen

this permits continued computation of the constant tuple arguments to ccall/cglobal/llvmcall with a linearized IR

vtjnash commented:
i decided to implement a few more optimizations on top of this: codegen should now capable of running the sysimg.jl code through to completion with the fully linear IR, with or without type-inference (as long as you don't try to generate Core.Inference)

vtjnash commented:
@JeffBezanson i don't know if you want issues opened against `*very-linear-mode*` yet, but there's a bug with the ccall `&` lowering since it is syntax and not a proper Expr node, but is getting moved out of argument position:
```
expand(quote function BigFloat()
               ccall((:mpfr_init2,:libmpfr), Void, (Ptr{BigFloat},), &z)
           end end)
---
        GenSym(9) = &((GenSym(5))(GenSym(8),GenSym(4)))
        return (top(ccall))((top(tuple))(:mpfr_init2,:libmpfr),Main.Void,(top(svec))((top(apply_type))(Main.Ptr,Main.BigFloat)),GenSym(9),GenSym(4))
    end))))), false))
```
